airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_incremental.py

```diff
@@ -170,70 +170,92 @@ async def test_read_sequential_slices(
+
+            # To avoid spamming APIs we only test a fraction of batches (4 or 5 states by default)
+            min_batches_to_test = 5
+            sample_rate = (len(unique_state_messages) // min_batches_to_test) or 1
+
+            mutating_stream_name_to_per_stream_state = dict()
+            for idx, state_message in enumerate(unique_state_messages):
+                assert state_message.type == Type.STATE
+
+                is_first_state_message = idx == 0
+                is_last_state_message = idx == len(unique_state_messages) - 1
+
+                if checkpoint_testing_strategy == CheckpointingStrategies.use_progression:
+                    # if first state message, skip
+                    # this is because we cannot assert if the first state message will result in new records
+                    # as in this case it is possible for a connector to return an empty state message when it first starts.
+                    # e.g. if the connector decides it wants to let the caller know that it has started with an empty state.
+                    if is_first_state_message:
+                        continue
+
+                    # if batching required, and not a sample, skip
+                    if idx % sample_rate != 0:
+                        continue
+
+                    # if last state message, skip
+                    # this is because we cannot assert if the last state message will result in new records
+                    # as in this case it is possible for a connector to return a previous state message.
+                    # e.g. if the connector is using pagination and the last page is only partially full
+                    if is_last_state_message:
+                        continue
+
+                state_input, mutating_stream_name_to_per_stream_state = self.get_next_state_input(
+                    state_message, mutating_stream_name_to_per_stream_state, is_per_stream
+                )
 
-            diff = naive_diff_records(records_1, records_N)
-            assert (
-                diff
-            ), f"Records for subsequent reads with new state should be different.\n\n records_1: {records_1} \n\n state: {state_input} \n\n records_{idx + 2}: {records_N} \n\n diff: {diff}"
+                output_N = await docker_runner.call_read_with_state(
+                    connector_config, configured_catalog_for_incremental_per_stream, state=state_input
+                )
+                records_N = filter_output(output_N, type_=Type.RECORD)
+
+                assert (
+                    # We assume that the output may be empty when we read the latest state, or it must produce some data if we are in the middle of our progression
+                    records_N
+                    or is_last_state_message
+                ), f"Read {idx + 2} of {len(unique_state_messages)} should produce at least one record.\n\n state: {state_input} \n\n records_{idx + 2}: {records_N}"
```

Comment:
Can we count the number of records in between state message from `output_1` and check that `records_N` matches this count?

Reply:
Yes we can, and we should also avoid using wrong calculation {idx + 2}. I am going to fix it as well